### PR TITLE
Adding a new setUserAsOwner flag to publish:gitlab action.

### PR DESCRIPTION
**Pull request by matteosilv**
Adding a new setUserAsOwner flag to publish:gitlab action. 

The field default is `false`. 

When true it will use the token configured in the gitlab integration for the matching host, to try and set the user logged in via `repoUrlPicker` `requestUserCredentials` OAuth flow as owner of the repository created in GitLab. 

This solves issues where the initial restrictions on the default branch does not allow a non-owner user to push the initial commit to create the default branch.

### CODE REVIEW

**Comment by matteosilv** [comment]
Do I just need a review right? It's my first pull and I could have missed something.

**Comment by awanlin** [comment]
> Do I just need a review right? It's my first pull and I could have missed something.

Yes, that's correct, one of the Backstage Maintainers will do a review and once that done will they will also merge it in. It's also time for Summer Vacation for the Backstage team so things might take a little longer. I'm just one of the many Community members who try and help out here and there.

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlab.ts
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlab.ts:175]: I'm not sure that we should be creating a new client here. We already have a token that we used further up, and that token was used to create the repo, so it should have the required scopes to update the owner.  I recommend we use the `client` already defined to further up.

**Review by benjdlambert — CHANGES_REQUESTED** [review]
Just a comment here about the client usage.

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlab.ts
Comment (benjdlambert) [reply plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlab.ts:175]: Probably don't need to check if there's an `integrationConfig.config.token` then either.

Comment (matteosilv) [reply plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlab.ts:175]: This workflow is supposed to work when the OAuth users do not have the rights to set themselves as owners of the repo. That's why I need to use the token in the integration config, because the users in our case were not able to push the initial protected branch nor setting themselves as owners.

Comment (matteosilv) [reply plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlab.ts:175]: To summarize:
- the user can create the repo and that's it. It cannot push on it. I know it's weird, but it's how gitlab works...
- only thing i could try to do is see, if I can use a different token for a single request, without creating a new client

Comment (benjdlambert) [reply plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlab.ts:175]: > the users can create the repo and that's it. They cannot push on it

this makes absolutely zero sense :joy: wasn't aware it worked like this, but ok, now I see the issue. 

Ok, let's go with this then, even it's a little unfortunate.

**Comment by benjdlambert** [comment]
Think you'll need to update the `api-reports` by running `yarn build:api-reports` and then commiting the changed files. 👍

**Comment by matteosilv** [comment]
Updated api-report.md

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlab.ts
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlab.ts:177]:
```suggestion
          token: integrationConfig.config.token,
```

Last nit I think!

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlab.ts
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlab.ts:174]: Maybe we can add a comment here above that describes the behaviour noted in the chat below and why the `integration.config` needs to be set etc.

**Review by benjdlambert — CHANGES_REQUESTED** [review]
Sorry, just two more small things I think

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlab.ts
Comment (matteosilv) [reply plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlab.ts:174]: Yeah sure, so that others can understand right away why it is needed

Comment (matteosilv) [reply plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlab.ts:180]: Just added the comment. Hope it is clear enough.

**Review by benjdlambert — APPROVED** [review]
Thanks again 👏

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlab.ts
Comment (dfl-aeb) [inline plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlab.ts:160]: The client resolution does only work, when you are authenticated via gitlab right? Cause we use an azure identity provider for login in backstage and a gitlab group access token to create resources in gitlab.
At the moment it uses the user_id of the gitlab group access token and sets it as owner of the repository.

Is there a way for us to set the owner the right way?

**Pull request merged**
